eGovPDX/portlandor
# [POWR-1782] - PL JS Cleanup - Drupal Behaviors attach in PL

## [POWR-1782 - PL JS Cleanup - Drupal Behaviors attach in PL](https://portland.atlassian.net/browse/POWR-1782)

This sets up the foundation well for JavaScript by:

- Breaking out independent functionality into it's own js file that's alongside it's scss file for ease of development on a single feature. The main `cloudy.js` just `import`'s each of those in. I did **not** change the contents of the JS functions for the files since git would have a hard time tracking those changes.
- Sets up WebPack to recognize that if we use `import Drupal from 'Drupal'` that we simply get that value from `window.Drupal` upon script load. This let's us use the modern `import` method for pulling in our dependencies instead of the immediate invoking function expression (aka `(function ($, Drupal) { // stuff })(jQuery, Drupal);` Also, when external dependencies and plugins are needed in future the `import` will allow WebPack to reduce it's file size better. This is set up in the `externals` config item in `webpack.config.js`. Done for jQuery too.
- Instead of importing the entirety of Bootstrap JS in, we have a single webpack entry (`bootstrap.js`) that in turn imports each piece of Bootstrap's JS (alerts, tooltips, etc). It still is including all of Bootstrap but this will be much easier to whittle down over time.
- Pattern Lab is set up to emulate the JS environment in Drupal. This is done by using `<script>` tags to pull in the same JS that Drupal does: same jQuery version and core plugins like "once", along with `drupal.js` and `domready` so we can use `Drupal.attachBehaviors` everywhere for consistency and it will fire the `attach` method inside when the page loads. This will help to keep the developer mental model consistent between the two environments and will reduce surprises.
- Setting the foundation in Pattern Lab for good docs for our patterns by including markdown parsing and syntax highlighting inside any `.js-markdown` element. This is **not** pulled into Drupal, only Pattern Lab.
- The main theme asset files of `main.js` and `style.scss` have been renamed to `cloudy.js` and `cloudy.scss` to align with Drupal standards and to provide better developer hints that these are coming from the `cloudy` theme. 
- Setting up asset compiling (both JS & Scss) to use the modern method of "transpire files based on their destined environment" i.e. tell it what browsers you support and it'll figure out what JS syntax needs changing and what CSS vendor prefixes are needed. This is set in the `.browserlistrc` file and is used by `@babel/preset-env` and `autoprefixer`. PS I set it to basically support IE11 and higher along with targets related to US since I doubt you get many non-US visitors. 
- Switched the production flag over to the standard method of setting the env var `NODE_ENV=production` since many minification plugins use that.

### Testing

- Run `npm install && npm start` in `cloudy` folder to install new dependencies, compile assets and start Pattern Lab.
- Open Pattern Lab and notice the lack of console errors
- Open up other Drupal pages where JS is firing and notice it working still

[POWR-1782]: https://portland.atlassian.net/browse/POWR-1782
